flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/co/CoBroadcastWithKeyedOperator.java

```diff
@@ -0,0 +1,323 @@
+			this.timerService = Preconditions.checkNotNull(timerService);
+		}
+
+		void setElement(StreamRecord<IN2> e) {
+			this.element = e;
+		}
+
+		@Override
+		public Long timestamp() {
+			checkState(element != null);
+			return element.getTimestamp();
+		}
+
+		@Override
+		public <K, V> ReadWriteBroadcastState<K, V> getBroadcastState(MapStateDescriptor<K, V> stateDescriptor) {
+			Preconditions.checkNotNull(stateDescriptor);
+			return  (ReadWriteBroadcastState<K, V>) states.get(stateDescriptor);
+		}
+
+		@Override
+		public <X> void output(OutputTag<X> outputTag, X value) {
+			checkArgument(outputTag != null, "OutputTag must not be null.");
+			output.collect(outputTag, new StreamRecord<>(value, element.getTimestamp()));
+		}
+
+		@Override
+		public long currentProcessingTime() {
+			return timerService.currentProcessingTime();
+		}
+
+		@Override
+		public long currentWatermark() {
+			return timerService.currentWatermark();
+		}
+
+		@Override
+		public <VS, S extends State> void applyToKeyedState(
+				final StateDescriptor<S, VS> stateDescriptor,
+				final KeyedStateFunction<KS, S> function) throws Exception {
+
+			keyedStateBackend.applyToAllKeys(
+					Preconditions.checkNotNull(stateDescriptor),
+					Preconditions.checkNotNull(function));
+		}
+	}
+
+	private class ReadOnlyContextImpl extends KeyedBroadcastProcessFunction<KS, IN1, IN2, OUT>.KeyedReadOnlyContext {
+
+		private final Map<MapStateDescriptor<?, ?>, ReadWriteBroadcastState<?, ?>> states;
+
+		private final TimerService timerService;
+
+		private StreamRecord<IN1> element;
+
+		ReadOnlyContextImpl(
+				final KeyedBroadcastProcessFunction<KS, IN1, IN2, OUT> function,
+				final Map<MapStateDescriptor<?, ?>, ReadWriteBroadcastState<?, ?>> broadcastStates,
+				final TimerService timerService) {
+
+			function.super();
+			this.states = Preconditions.checkNotNull(broadcastStates);
+			this.timerService = Preconditions.checkNotNull(timerService);
+		}
+
+		void setElement(StreamRecord<IN1> e) {
+			this.element = e;
+		}
+
+		@Override
+		public Long timestamp() {
+			checkState(element != null);
+			return element.hasTimestamp() ? element.getTimestamp() : null;
+		}
+
+		@Override
+		public TimerService timerService() {
+			return timerService;
+		}
+
+		@Override
+		public long currentProcessingTime() {
+			return timerService.currentProcessingTime();
+		}
+
+		@Override
+		public long currentWatermark() {
+			return timerService.currentWatermark();
+		}
+
+		@Override
+		public <X> void output(OutputTag<X> outputTag, X value) {
+			checkArgument(outputTag != null, "OutputTag must not be null.");
+			output.collect(outputTag, new StreamRecord<>(value, element.getTimestamp()));
+		}
+
+		@Override
+		public  <K, V> ReadOnlyBroadcastState<K, V> getBroadcastState(MapStateDescriptor<K, V> stateDescriptor) {
+			Preconditions.checkNotNull(stateDescriptor);
+			return  (ReadOnlyBroadcastState<K, V>) states.get(stateDescriptor);
```

Comment:
I think we should throw an exception here that explains what's going on if a state is not available.